### DISCUSSION TITLE
How can I build EGSnrc under Windows?

### DISCUSSION THREAD

**Discussion by 1035326373** [nrc-cnrc/EGSnrc#767] [Questions] [answered]
Building projects with visual studio: I want to learn the details of the code and do some secondary development. How can I build the project under windows? Is there a tool similar to `cmake` available?

**Comment by rtownson** [comment, marked as answer]
We don't use cmake. On linux there is a configuration script that does it all - on windows it's really just the installer gui that is comprehensive. But for development essentially if you make a change in egs++ then in HEN_HOUSE/egs++ do `mingw32-make` and it will handle the whole egs++ library. If you change the mortran code or an application, then you just need to recompile the application (e.g. in egs_home/egs_app or whichever directory do `mingw32-make`).

**Reply by 1035326373** [reply]
Although it is convenient to recompile code on Windows platform. However, if there is an error in the part of the changed code, it will be difficult to find the error. If there is a good way to establish the project with visual studio, it will be more convenient for beginners to learn the code structure and do secondary development.

**Reply by mchamberland** [reply]
Not sure specifically about Windows, but on a Mac, if you compile EGSnrc with the debugging option, you can open up the source code in Xcode (Mac IDE) and tell Xcode which executable to run when testing the code. Then you get full debugging for the C++ part of the code. Not sure about the FORTRAN though. 

So, for Visual Studio, I would try something similar. Build the code with debugging option first, then load the source code in VS, then try to find a way to point it to the executable to run. 

Or the other EGSperts can share how they set up their environment for debugging. 😅

**Reply by crcrewso** [reply]
> Although it is convenient to recompile code on Windows platform. However, if there is an error in the part of the changed code, it will be difficult to find the error. If there is a good way to establish the project with visual studio, it will be more convenient for beginners to learn the code structure and do secondary development.

I've thought a lot about this. With the current use cases, and tools I would advise using WSL and follow the linux install procedure. You are able to use VS Code, installed on windows, to work with, and debug the linux version. In my opinion this workflow is superior to mingw in all ways.

**Reply by 1035326373** [reply]
I think EGSnrc is still not good enough in cross platform. It is necessary to use some good tools to reduce the threshold for new users to join.

**Reply by ftessier** [reply]
That is what we are doing: using GNU/Linux 🤡  Sorry, could not resist! 🤓

**Reply by ftessier** [reply]
@crcrewso Have you had a chance to look into WSL integration in Windows 11?

**Reply by Kelvin-Ngowi** [reply]
this is what i am trying to do at the moment

**Reply by crcrewso** [reply]
> this is what i am trying to do at the moment

Microsoft finally marked WSL as released. If you can find it in the windows store, then
- Install the app from the store
- Launch the new Ubuntu terminal
- Follow the instructions provided for Linux
- You may have to install make, cmake, gfortran, gcc, tcl, tk, and grace 
  - `sudo apt install make cmake gfortran gcc tcl-dev tk-dev`
- From there you should be able to use a text editor like VS code to edit the source files,

The recently released version 1.0 of WSL does a lot of the hard work for free.

**Comment by crcrewso** [comment]
Not yet. But as soon as windows 11 hits general release it will be on my list ( I don’t have a spare machine to install the beta on currently).

**Reply by crcrewso** [reply]
@ftessier 
> @crcrewso Have you had a chance to look into WSL integration in Windows 11?

I have now, it's really easy. I was in the middle of writing up a how-to when Microsoft decided to put WSL on the Microsoft Store. I'll finish writing that up soon.
[Announcement Here](https://devblogs.microsoft.com/commandline/a-preview-of-wsl-in-the-microsoft-store-is-now-available/)
